docs/spec_v6.md

```diff
@@ -0,0 +1,220 @@
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+
+### URLs
+
+In [BNF syntax](https://tools.ietf.org/html/rfc5234), Twirp's URLs
+have the following format:
+
+```bnf
+URL ::= Base-URL "/" [ Package "/" ] Service "/" Method
+```
+
+The Twirp wire protocol uses HTTP URLs to specify the RPC
+endpoints on the server for sending the requests. Such direct mapping
+makes the request routing simple and efficient. The Twirp URLs have
+the following components.
+
+* **Base-URL** is the virtual location of a Twirp API server, which is typically
+  published via API documentation or service discovery. It should be a RFC 3986
+  URL. In the language of
+  [RFC 3986 Section 3](https://tools.ietf.org/html/rfc3986?#section-3), it must
+  have scheme and authority components. It may have a path component. It must
+  not have query or fragment components.
+* **Package** is the proto `package` name for an API, which is often
+  considered as an API version. For example,
+  `example.calendar.v1`. This component is omitted if the API
+  definition doesn't have a package name.
+* **Service** is the proto `service` name for an API. For example,
+  `CalendarService`.
+* **Method** is the proto `rpc` name for an API method. For example,
+  `CreateEvent`.
+
+### Requests
+
+Twirp always uses HTTP POST method to send requests, because it
+closely matches the semantics of RPC methods.
+
+The **Request-Headers** are normal HTTP headers. The Twirp wire
+protocol uses the following headers.
+
+* **Content-Type** header indicates the proto message encoding, which
+  should be one of "application/protobuf", "application/json". The
+  server uses this value to decide how to parse the request body,
+  and encode the response body.
+
+The **Request-Body** is the encoded request message, contained in the
+HTTP request body. The encoding is specified by the `Content-Type`
+header.
+
+### Responses
+
+The **Response-Headers** are just normal HTTP response headers. The
+Twirp wire protocol uses the following headers.
+
+* **Content-Type** The value should be either "application/protobuf"
+  or "application/json" to indicate the encoding of the response
+  message. It must match the "Content-Type" header in the request.
+
+The **Request-Body** is the encoded response message contained in the
+HTTP response body. The encoding is specified by the `Content-Type`
+header.
+
+### Example
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com/userprefix".
+
+```proto
+syntax = "proto3";
+
+package example.echoer;
+
+service Echo {
+  rpc Hello(HelloRequest) returns (HelloResponse);
+}
+
+message HelloRequest {
+  string message;
+}
+
+message HelloResponse {
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /userprefix/example.echoer/Echo/Hello HTTP/1.1
```

Comment:
Using `/` as a delimiter forces us to deal with cases of overlap. Consider the following two:

One has no package declaration, and describes a service called "Foo" with a method "Hello". It gets requests like `POST /userprefix/Foo/Hello`.

The other has a package declaration of "package Foo;" .. not matching the normal lower-case pattern, but a package declaration nonetheless. It describes some service, maybe "Bar" with a method "Goodbye". It gets requests like `POST /userprefix/Foo/Bar/Goodbye`.

How does a user route requests to those two services? Do they need to count the slashes? Peeking into the later parts of the URI path and only partially understanding them is an invitation for maintenance and security problems down the road.

Reply:
Oof, this is an extremely good point, and might torpedo the `/` idea here. I'm going to remove the `.` -> `/` change from this commit.

Let's discuss in #86.